### PR TITLE
Protect Shields migration from unregistered preferences

### PR DESCRIPTION
**Pull request by mariospr**
Resolves https://github.com/brave/brave-browser/issues/13508

## Submitter Checklist:

- [X] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue.
- [X] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [X] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [X] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [X] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [X] Ran `git rebase master` (if needed).
- [ ] Requested a security/privacy review as needed.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc
Comment (iefremov) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.cc:257]: mb we should check `FindPreference` as well? Here is a comment for `HasPrefPath` (the crash was caused specifically by an unregistered preference)
```
  // Returns true if a value has been set for the specified path.
  // NOTE: this is NOT the same as FindPreference. In particular
  // FindPreference returns whether RegisterXXX has been invoked, where as
  // this checks if a value exists for the path.
  ```

Comment (mariospr) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.cc:257]: I'm not sure whether that comment might actually be stale since the implementation of `HasPrefPath()` relies on `FindPreference()` after all:
```
bool PrefService::HasPrefPath(const std::string& path) const {
  const Preference* pref = FindPreference(path);
  return pref && !pref->IsDefaultValue();
}
```

Also, I can confirm that the newly added unit test crashes with the error below if I remove this early return:
```
[1273689:1273689:0112/173326.587131:68451040682:FATAL:pref_service.cc(546)] Check failed: false. Trying to get an unregistered pref: profile.content_settings.exceptions.unknown_setting
```

So it seems to me that `HasPrefPath()` does actually do the right thing after all, WDYT?

Comment (bridiver) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.cc:257]: I think it would be better to have a check against the expected resource ids in MigrateShieldsSettingsFromResourceIds instead of trying to filter out things later. There is a small and specific list of things we want to migrate and we shouldn't be passing anything else through

Comment (mariospr) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.cc:257]: Sounds good to me too, actually I initially had the check in that method instead of this one, but then I thought it would be slightly better to have it inside `MigrateShieldsSettingsFromResourceIdsForOneType()` so that, no matter how we reached there, we're always protected.

No strong opinion though, and I also get the point of having that check in the main migration method along with other checks (e.g. early filtering out "fingerprinting"), will change the PR to reflect that.

Comment (mariospr) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: @bridiver I just force-pushed an updated PR now that drops that early return in favour of a specific check in the other method that filters out anything that does not match a Shields-related ContentSettingsType name, which also means we can drop the specific check on `brave_shields::kObsoleteFingerprinting` that @iefremov added previously.

PTAL, thanks!

File: components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc
Comment (iefremov) [inline components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc:74]: const int& -> int

File: components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc
Comment (iefremov) [inline components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc:520]: is it really needed?

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc
Comment (iefremov) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: we can add a DCHECK to `MigrateShieldsSettingsFromResourceIdsForOneType`

**Review by iefremov — APPROVED** [review]

File: components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc
Comment (mariospr) [reply components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc:520]: I think this is probable a better question for @bridiver, I just put it here following the lead of the other tests that existed in this file before I added these newer tests for the migration from ResourceIdentifiers, that is: `TestShieldsSettingsMigration` and `TestShieldsSettingsMigrationVersion`.

File: components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc
Comment (mariospr) [reply components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc:74]: Oops! It's actually in many places in this file, fixed all of them.

**Comment by mariospr** [comment]
> mariospr force-pushed the issues/13508 branch from 0052ab8 to da306b1 16 minutes ago

Just clarifying that the only change in this force-push is the removal of a call to `CopyPluginSettingsForMigration()` from the test added in this PR, since that's no longer needed thanks to https://github.com/brave/brave-core/pull/7575.

@bridiver Waiting for your final approval after Ivan's LGTM, thanks in advance

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc
Comment (mariospr) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: Added

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc
Comment (bridiver) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.cc:272]: shouldn't this be a `CHECK` with supporting text explanation since we know this will cause a crash?

File: components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc
Comment (bridiver) [inline components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc:97]: This is a confusing method name and either needs some comments or a rename. Why would we initialize invalid settings? I assume that isn't what is happening here, but I'm having trouble following what is happening

File: components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc
Comment (bridiver) [reply components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc:518]: @iefremov I think the test might crash without it, but it's been a while so I'm not sure

Comment (bridiver) [reply components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc:518]: the chromium tests do the same thing

**Review by bridiver — APPROVED** [review]

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc
Comment (mariospr) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: True, fixed

File: components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc
Comment (mariospr) [reply components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc:97]: Based on our conversation on Slack, I put some explanatory comments in the test where this is used from to make it more understandable.

Comment (mariospr) [reply components/content_settings/core/browser/brave_content_settings_pref_provider_unittest.cc]: I also renamed the method to `InitializeUnsupportedShieldSettingInDictionary()`, to make it clear that it's not that we're initializing invalid data, but simply something that is not supported.

**Comment by mariospr** [comment]
All checks have finally passed (for some reason the macos job took a long time to start, just finished a couple of hours ago), so I just merged it to master now.

**Comment by kjozwiak** [comment]
Verification **PASSED** on `Win 10 x64` using the following build:

```
Brave | 1.21.5 Chromium: 88.0.4324.87 (Official Build) nightly (64-bit)
-- | --
Revision | dd01ff8f58c65af81127ad5c105c79d5b571d8f3-refs/branch-heads/4324@{#1702}
OS | Windows 10 OS Version 2009 (Build 19042.746)
```

* download https://github.com/brave/brave-browser/releases/v1.21.3
* replace the `Preference` file from `Brave-Browser-Nightly\User Data\Default` using [Preferences.txt](https://github.com/brave/brave-core/files/5822406/Preferences.txt) (remove `.txt`)
* launch `1.21.3 CR: 88.0.4324.87` and you'll notice that Nightly doesn't start (crashes on startup)
* update/install `1.21.5 CR: 88.0.4324.87` and relaunch

**Pull request merged**
